lib/temporalio/worker/workflow_worker.rb

```diff
@@ -0,0 +1,127 @@
+require 'temporal/sdk/core/workflow_commands/workflow_commands_pb'
+require 'temporalio/error/failure'
+require 'temporalio/errors'
+require 'temporalio/interceptor'
+require 'temporalio/interceptor/chain'
+require 'temporalio/worker/workflow_runner'
+
+module Temporalio
+  class Worker
+    # @api private
+    class WorkflowWorker
+      def initialize(namespace, task_queue, worker, workflows, converter, interceptors)
+        @namespace = namespace
+        @task_queue = task_queue
+        @worker = worker
+        @workflows = prepare_workflows(workflows)
+        @converter = converter
+        @inbound_interceptors = Temporalio::Interceptor::Chain.new(
+          Temporalio::Interceptor.filter(interceptors, :workflow_inbound),
+        )
+        @outbound_interceptors = Temporalio::Interceptor::Chain.new(
+          Temporalio::Interceptor.filter(interceptors, :workflow_outbound),
+        )
+        @drain_queue = Queue.new
+        @running_workflows = {}
+      end
+
+      def run(reactor)
+        # @type var outstanding_tasks: Array[Async::Task]
+        outstanding_tasks = []
+
+        loop do
+          workflow_activation = worker.poll_workflow_activation
+          outstanding_tasks << reactor.async do |async_task|
+            handle_activation(workflow_activation)
+          ensure
+            outstanding_tasks.delete(async_task)
+          end
+        end
+      rescue Temporalio::Bridge::Error::WorkerShutdown
+        # No need to re-raise this error, it's a part of a normal shutdown
+      ensure
+        outstanding_tasks.each(&:wait)
+        drain_queue.close
+      end
+
+      def drain
+        drain_queue.pop
```

Comment:
Can you please clarify the purpose for the `drain_queue` and this `drain` method.

Also, I'm wondering:
- why is this a queue... wouldn't a mutex make more sense here?
- why it is exposed through `attr_reader` below?